README
NAME
    Net::Connection::Sniffer -- gather stats on network connections

SYNOPSIS
      use Net::Connection::Sniffer;

      gather($config)

DESCRIPTION
    Net::Connection::Sniffer is a perl module to gather connection
    statistics by listening to ethernet traffic. Traffic is filtered using
    standard BPF notation as described in the "tcpdump" documentation and
    implemented using the standard "pcap" library to sniff packets on host
    network interfaces.

CONFIGURATION
    Create a directory with appropriate permissions for the pid file and the
    profile statistics dump file. Typical installation:

      mkdir -p /var/run/nc.sniffer

    Edit the nc.sniffer.pl.sample file to change or set the following:

      my $config = {

      # specify the directory for the pid file for this daemon.
      # The directory must exist and have writable permissions.
      # [required]
      #
            piddir  =>  '/var/run/nc.sniffer',

      # specify the directory for the statistics file for this 
      # daemon. The directory must exist and have writable
      # permissions
      # [required]
      #
            sniffer =>  '/var/run/nc.sniffer',

      # BPF filter statement. See examples below.
      # [required]
      #
            bpf     => 'src host myhost.com and tcp port 80',

      # size of the portion of packet to capture, defaults
      # to the minimum size necessary to determine the
      # source and destination IP addresses and port numbers
      # [optional]          ETH_head + IPV4_head + 4

      #     snaplen => 38,

      # filter condition: payload must contain this string.
      # case insensitive match of the payload data to this string. 
      # [optional]

      #     match   => 'somestring',

      # filter condition: payload must NOT contain this string.
      # case insensitive match of the payload data to this string.
      # [optional]

      #     nomatch => 'some.other.string',

      # offset of the payload from the packet start
      # typically at least 60 for tcp, 44 for udp
      # [optional]... but [required] for 'match', 'nomatch'
      #
      #     payload => 44,

      # UDP listen port to trigger a dump file
      # [optional]
      #
            port    => 10004,

      # HOST address on which to listen for dump request
      # may be one of a HOSTNAME, IP address, or
      # strings 'INADDR_ANY', 'INADDR_LOOPBACK'
      # [optional] default 127.0.0.1 == INADDR_LOOPBACK
      #
            host    => 'INADDR_LOOPBACK',

      # ALLOWED connecting host(s)
      # may be HOSTNAME or IP address
      # [optional] default 127.0.0.1
      #
            allowed => ['127.0.0.1',],

      };

    To generate a web report to STDOUT with or without a cache file, edit
    the nc.sniffer.cgi.sample file to change or set the configuration
    parameters. See the Net::Connection::Sniffer::Report manpage::web_report
    or the sample file for details.

      Usage: <!--#exec cmd="./nc.sniffer.cgi 0" -->
        or   <!--#exec cmd="./nc.sniffer.cgi 1" -->

    where an argument of "0" produces a report ordered by /24 by usage and
    an argument of "1" produces a report ordered by subdomain by usage.

REMOTE CONFIGURATION, multiple hosts
    To configure the reporting function to retrieve statistics from multiple
    remote hosts (and localhost) do the following:

            1) read the config section of
               nc.sniffer.coalesce.cgi.sample
            2) read the config section of
               nc.sniffer.dump.pl.sample 

    On the remote host(s), install nc.sniffer.dump.pl in an appropriate
    sandbox account and install an ssh certificate to permit access to the
    sandbox ssh executable as well as the directory from which to rsync the
    stats file on that host.

    nc.sniffer.dump.pl should be installed mode 755 or as appropriate to be
    accessed remotely by the ssh -e function.

    On the web host, configure nc.sniffer.coalesce.cgi and place the
    execution cgi string in your web page to produce the report

    nc.sniffer.coalesce.cgi should be SUID to the web user, not root, so
    that the web engine can safely execute the script. The ssh certificate
    must be generated for the web user and go in the nobody:nogroup/.ssh
    directory (or equivalent web user directory).

      usage: <!--#exec cmd="./nc.sniffer.coalesce.cgi" -->

OPERATION
    Launch the daemon with the command:

            nc.sniffer.pl start

      Syntax: nc.sniffer.pl start
              nc.sniffer.pl stop
              nc.sniffer.pl restart
              nc.sniffer.pl status
              nc.sniffer.pl dump
              nc.sniffer.pl config

              -d switch may be added to
               redirect output to STDERR

    On most systems it will be necessary to wrap a shell script around
    nc.sniffer.pl if the path for perl is not in scope during boot.

      #!/bin/sh
      #
      # shell script 'rc.nc.sniffer'
      #
      /path/to/nc.sniffer.pl $*

    A sample shell script is included in the distribution as rc.nc.sniffer

    To run multiple copies of nc.sniffer for data collection on various
    ports or IP's at the same time, name them:

            nc.sniffer1.pl
            nc.sniffer2.pl
            etc...

      start         start daemon if not running, write pid file
      stop          stop a running daemon
      restart       do stop, then start
      status        report if daemon running or not
      dump          refresh/write statistics file
      config        print configuration to STDOUT

SIGNALS
    The statistics information will be written to the file specified in the
    configuration upon receipt of a SIG USR1

            SIG     TERM            write stats file, terminate
            SIG     HUP             write stats file, start over
            SIG     USR1            write statistics file

UDP listener -- statistics file dump
    If the nc.sniffer daemon is configured for a UDP listen port, sending a
    message dump will produce the same result as SIG USR1. The daemon will
    respond OK timestamp, but this is NOT syncronized with the file dump and
    only indicates that the statistics file should not have a timestamp
    earlier that the epoch value returned. When either a dump or SIG USR1 is
    issued, you must check the ctime of the file to determine if it has been
    updated.

BUGS / RESTRICTIONS
    Net::Connection::Sniffer uses libpcap. The data collection is
    accomplished using a selectable capture device which is NOT SUPPORTED on
    Windows and some older BSD platforms. The next two paragraphs are from
    the pcap library and describe the platform limitations.

    Some "BPF" ...devices do not support select() or poll() (for example,
    regular network devices on FreeBSD 4.3 and 4.4, and Endace DAG
    devices)...

    ...On most versions of most BSDs (including Mac OS X), select() and
    poll() do not work correctly on BPF devices. "While a BPF file
    descriptor will be returned" ...on most of those versions (the
    exceptions being FreeBSD 4.3 and 4.4), a simple select() or poll() will
    not return even after a... "specified timeout" expires... ...In FreeBSD
    4.6 and later, select() and poll() work correctly on BPF devices...

EXAMPLES
  BPF examples

    The bpf entry in the configuration hash uses the standard language
    documented in detail in the tcpdump man(1) page. The bpf statement must
    contain at a minimum, 'host somename [or IP address]'. The host
    specification must be for a single unique IP address and be the first
    such specification if there are multiple src/dest host specifications in
    the statment.

    Capture all traffic to/from a particular host:

      bpf   => 'host particular.host.com',

    Capture traffic to/from your mail server:

      bpf   => 'host my.mx.com and tcp port 25',

    Capture request traffic arriving at your DNS server:

      bpf   => 'dst host my.dns.com and udp port 53',

    Capture response traffic leaving your DNS server:

      bpf   => 'src host my.dns.com and udp port 53',

  Content MATCH/NOMATCH examples

    The match and nomatch configuration entries can be used to further
    discriminate which packets to sniff. When the match entry is set, only
    packets which meet the BPF criteria AND have matching data within the
    packet capture buffer are selected for analysis. Conversely, when the
    nomatch entry is set, packets which meet the BPF criteria and match the
    nomatch string are unconditionally dropped. match and nomatch may both
    be set.

    NOTE: that matches are made on a case insensitive basis.

    Capture request traffic arriving at the DNS port with a query for
    "somedomain.com". From RFC1035, we know that a datagram might need to
    use the domain names F.ISI.ARPA, FOO.F.ISI.ARPA, ARPA, and the root.
    Ignoring the other fields of the message, these domain names might be
    represented as:

           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
        20 |      decimal 1        |           F           |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
        22 |      decimal 3        |           I           |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
        24 |           S           |           I           |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
        26 |      decimal 4        |           A           |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
        28 |           R           |           P           |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
        30 |           A           |           0           |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+

           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
        40 |      decmial 3        |           F           |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
        42 |           O           |           O           |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
        44 | 1  1|            decimal 20                   |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+

           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
        64 | 1  1|            decimal 26                   |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+

           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
        92 |      decimal 0        |                       |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+

    Our examples somedomain.com would be represented in the datagram as
    follows:

           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
        20 |      decimal 10       |           s           |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
        22 |           o           |           m           |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
        24 |           e           |           d           |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
        26 |           o           |           m           |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
        28 |           a           |           i           |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
        30 |           n           |      decimal 3        |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+             
        30 |           c           |           o           |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+             
        30 |           m           |      decimal 0        |
           +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+             

    This translates to the perl string:

            where 10 becomes hex \x{a}

      $string = "\x{a}somedomain\x{3}com"

    The offset of the query QUESTION is:

      ETH header    16
      IP header     20
      UDP header    8
      Query head    12
                    --
                    54

    and the snaplen needs to be long enough to alway capture the domain
    name. There, our example configuration becomes:

      bpf     => 'dst host my.dns.com and udp port 53',
      match   => "\x{a}somedomain\x{3}com",
      snaplen => 90,
            # eth head + ip head + udp head + query head
      payload => 54,

DUMP FILE FORMAT
    The dump file is written in a format compatible with that produced by
    Data::Dumper. It may be imported for analysis using Perl's 'do' or by
    using File::SafeDO.

      # start:        1145923212,     Mon Apr 24 17:00:12 2006
      # current:      1145923334,     Mon Apr 24 17:02:14 2006
      # hits:         3832 per minute
      # bytes:        5927 per second
      # users:        1234 users now
      # device:       eth1:1  non-promiscuous
      # bpf:          dst host my.host.com
      # [optional if match/nomatch present]
      # fragment:   nn -- mm
      # contains:   match.string
      # excludes:   nomatch.string
      {
        my $dump = {
           '69.3.95.131'     => {
                    B       => 240,
                    C       => 4,
                    E       => 1145760699,
                    N       => ['hostname1','hostname2','...'],
                    R       => 723,
                    S       => 1145757331,
                    T       => 1145790478,
                    W       => 43359,
            },
      }

    * start:
        The start time of this data collection in seconds since the epoch
        and local time.

    * current:
        The time the file was written in seconds since the epoch and local
        time.

    * hits:
        The connections per minute collected by this filter configuration.

    * bytes:
        The bandwidth in bytes per second collected by this filter
        configuration.

    * users:
        The total number of discreet hosts logged at this instant

    * device:
        The network device being sniffed and whether or not the device is in
        promiscuous mode.

    * bpf:
        The bpf statment used for data collection

    * value hash pointer for one or more IP addresses.
        Time values are seconds since the epoch.

          Hash pointer = {

              IP address => {
                  B     => incremental byte count
                  C     => incremental connection count
                  E     => last update time
                  N     => ['hostname1','hostname2','...'],
                  R     => connections / hour
                  S     => start time this data set
                  T     => TTL expiration for hostname
                  W     => bytes / hour
              },

              next IP address => {
                ...

        NOTE: if the hostname lookup results in an NXDOMAIN return, the
        hostname will be parsed from the SOA record and presented prefixed
        with a colon

          i.e.  ':soahost.com'

EXPORTS
    Only one function is exported by Sniffer.pm. This function is called in
    the nc.sniffer.pl.sample script to launch the nc.sniffer daemon.

    gather($config);
        Launch the nc.sniffer daemon.

          input:        config hash
          returns:      nothing (exits)

PREREQUISITES
    The "pcap" library ("libpcap") which is part of "tcpdump" and is
    included in most *nix distributions. Available from:

      http://sourceforge.net/projects/libpcap/

    the NetAddr::IP::Util manpage which is part of distribution the
    NetAddr::IP manpage

    the Net::Interface manpage

    the Net::DNS::Codes manpage

    the Net::DNS::ToolKit manpage

    the Net::NBsocket manpage

    the Proc::PidUtil manpage

    the Sys::Hostname::FQDN manpage

    the Sys::Sig manpage

BUGS
    There is a memory leak when run under Perl 5.0503 that has not yielded
    to debug attempts. This leak is not present in Perl 5.0601. Not tested
    in other versions. From reading through the Changes file for the
    transition between versions 5.005 and 5.6, I'm reasonably sure it is a
    scalar leak in Perl itself that was corrected with the updates to 5.6.

    My recommend fix for now when running with Perl versions older than 5.6
    is to restart the daemon daily to prevent excessive memory consumption.

COPYRIGHT 2004 - 2011
    Michael Robinton <michael@bizsystems.com>

    All rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either:

      a) the GNU General Public License as published by the Free
      Software Foundation; either version 2, or (at your option) any
      later version, or

      b) the "Artistic License" which comes with this distribution.

    This program is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See either the GNU
    General Public License or the Artistic License for more details.

    You should have received a copy of the Artistic License with this
    distribution, in the file named "Artistic". If not, I'll be glad to
    provide one.

    You should also have received a copy of the GNU General Public License
    along with this program in the file named "Copying". If not, write to
    the

            Free Software Foundation, Inc.
            59 Temple Place, Suite 330
            Boston, MA  02111-1307, USA

    or visit their web page on the internet at:

            http://www.gnu.org/copyleft/gpl.html.

AUTHOR
    Michael Robinton <michael@bizsystems.com>

SEE ALSO
            man (1) tcpdump
            man (3) pcap

    the Net::Connection::Sniffer::Report manpage



Changes
Revision history for Perl extension Net::Connection::Sniffer

0.35  Wed Jun  5 14:08:42 PDT 2013
	in Report v0.13
	add chk_lock to test for other processes running coalesce
	add load average testing to reduce CPU load to 70% max

0.34  Sun Sep 25 17:43:53 PDT 2011
	this time "really" fix the typo in dyn_bind documentation
	Net::Connection::Sniffer::Report v0.12

0.33  Sun Sep 25 17:29:13 PDT 2011
	update type on dyn_bind documentation in
	Net::Connection::Sniffer::Report v0.11

0.32  Sun Sep 25 16:16:37 PDT 2011
	removed dyn_bind from Net::Connection::Sniffer::Report v0.10

0.31  Tue May 26 16:15:28 PDT 2009
	add contrib/chkIFrunning.pl
	a script to check if sniffer is running and restart if necessary

0.30  Sat Mar  7 12:11:24 PST 2009
	add compiler and linker flags to environment for various
	shared "flags" and hosts that put "libpcap" in /usr/local
	instead of /usr

	clean up "char/u_char" casting differences in Sniffer.xs
	for picky Solaris compiler

0.29  Wed Feb 25 21:44:14 PST 2009
	update get_if() to use Net::Interface 1.00x

	remove early "use Devel::CheckLib" so that automated
	test routines have a chance to find in the the Makefile.pl
	prerequisites list.

	add definitions of ETH_HLEN and IP_HDR
	for systems that don't provide it

	add "configure" script

	remove dependency of Devel::CheckLib

0.28  NOT RELEASED
	correct dependencies in Makefile.PL

	kill small memory leak

0.27  Wed Oct 15 14:19:29 PDT 2008
	include Devel::CheckLib version level
	require Net::DNS::ToolKit verison 0.41 or higher

0.26  Mon Oct  6 14:02:13 PDT 2008
	include Devel::CheckLib in prereques

0.25  Thu Oct  2 13:39:49 PDT 2008
	add CPAN automated tester cutout if pcap lib is not installed

0.24  Mon Jan 14 11:48:04 PST 2008
	force timecheck every second, even if very busy
	so that old IP's will expire

0.23  Wed Apr 25 16:52:52 PDT 2007
	add redundant DNS name reporting to compensate for occasional
	missing rDNS entries for remote hosts

0.22  Mon Apr  2 18:14:04 PDT 2007
	add SOA lookup of hostname where DNS query result is NXDOMAIN

0.21  Thu Dec 21 17:41:45 PST 2006
	update documentation and report formatting in Report.pm v0.08

0.20  Thu Dec 21 15:25:51 PST 2006
	fix problem in Report.pm v0.07 where localhost is not updated
	with a rem_update

0.19  Wed Dec 20 18:10:49 PST 2006
	NOT released

	add remote statistics collection from remote hosts
	with coalesced statistic reporting to 
	Report.pm v0.06

	Update documentation to reflect above.
	add nc.sniffer.coalesc.cgi.sample
	add nc.sniffer.dump.pl.sample

0.18  Fri Oct 27 11:53:34 PDT 2006
	add local $^W = 1; # use warnings
	to nc.sniffer

	add BUGS memo about known memory leak when
	running under Perl 5.00503

0.17  Sun Sep 10 13:42:39 PDT 2006
	fix 'config' printout of sub check_run in
	Sniffer.pm which did not expand array ref

0.16  Sun Sep 10 03:09:29 PDT 2006
	rename to get PAUSE to upload again

0.15  Wed Sep  6 12:38:54 PDT 2006
	add payload string match capability

	add perl 5.0503 compability updates

0.14  Sat Aug 19 14:01:08 PDT 2006
	add test for zero length stat file in Report.pm v0.04

	NetAddr::IP::Lite 0.06 / NetAddr::IP::Util 0.11
	required for operation on Sparc

0.13  Sat Jun  3 12:48:26 PDT 2006
	add 'sub equal' to Util.pm v0.02

	update Report v0.03 'by_cidr' so that it
	does not truncate a single or last entry
	from the stats hash

0.12  Fri Jun  2 13:02:48 PDT 2006
	Oops! missed another makefile prequisite entry

0.11  Fri Jun  2 12:46:38 PDT 2006
	Left prequisite Net::Interface
	out of Makefile.PL

0.10  Fri Jun  2 11:18:38 PDT 2006
	Documentation update

0.09  Thu Jun  1 10:38:33 PDT 2006
	first release to CPAN

	clean up documentation

0.08  Wed May 31 17:45:13 PDT 2006
	optimizations to reduce cpu cycles

0.07  Tue May 30 19:06:03 PDT 2006
	remove unused vars in Perl and C that are
	pushed back and forth between xs <-> perl

	change "B" accumulator from UV to NV

	separate double / u_int32_t arithmetic into
	individual pieces -- compiler appears to mis-align
	variable addresses with with no warnings

0.06  Mon May 29 13:16:44 PDT 2006
	release of port to 'C'

0.05  this and previous versions not released

0.04  Fri May 19 18:08:00 PDT 2006
	replace missing do_while in daemon v03

0.03  Wed May 17 18:58:07 PDT 2006
	further refinement ahead of conversion

	deployed to test with bug

0.02  Tue May 16 14:48:58 PDT 2006
	rearrange stuff ahead of conversion of core to 'C'

0.01  Sat May 13 19:30:43 PDT 2006
	initial release


lib/Net/Connection/Sniffer/Util.pm
#!/usr/bin/perl
package Net::Connection::Sniffer::Util;

use strict;

use NetAddr::IP::Util qw(
	sub128
	ipanyto6
	ipv6_aton
);
use vars qw($VERSION);

$VERSION = do { my @r = (q$Revision: 0.01 $ =~ /\d+/g); sprintf "%d."."%02d" x $#r, @r };

=head1 NAME

Net::Connection::Sniffer::Util -- netaddr utilities

=head1 SYNOPSIS

  use Net::Connection::Sniffer::Util;

  my $ip = newcidr24 Net::Connection::Sniffer::Util($netaddr);

  my $ipcopy = $ip->copy;

  if ($someip->within($ip)) {
	do something...

  if ($ip->contains($someip)) {
	do something...

=head1 DESCRIPTION

B<Net::Connection::Sniffer::Util> is a lite weight perl module to do NetAddr::IP like 
operations on ip addresses in either 32 or 128 bit formats.

=over 4

=item * my $ip = newcidr24;

Create a blessed address object with cidr/24 network/broadcast bounds (assumes 32
bit addressing), for ipv6, this will be cidr/120

=cut

my $oxFF	= ipv6_aton('::FF');
my $oxFF00	= ~ $oxFF;

sub newcidr24 {
  my($proto,$naddr) = @_;
  my $class = ref($proto) || $proto;
  my $self  = {};
  $self->{IP} = $naddr = ipanyto6($naddr);
  $self->{BC} = $naddr | $oxFF;
  $self->{NT} = $naddr & $oxFF00;
  bless ($self, $class);
  return $self;
}


=item * my $ipcopy = $ip->copy;

Copy a blessed network address object to a new blessed object;

Returns 'undef' if $ip is not a Net::Connection::Sniffer::Util object.

=cut

sub copy {
  my $proto = shift;
  return undef unless (my $class = ref($proto)) eq __PACKAGE__;
  my $self = {};
  @{$self}{qw(IP BC NT)} = @{$proto}{qw(IP BC NT)};
  bless ($self,$class);
  return $self;
}

=item * $rv = $someip->within($ip);

Check to see if $someip is within the cidr of $ip. i.e. 

  network address <= $someip <= broadcast address

  input:	ip object for range check
  returns:	true if within, else false

Returns 'undef' if $someip and $ip are not Net::Connection::Sniffer::Util objects.

=cut

sub within {
  my($sref,$ip) = @_;
  return undef unless 	ref($sref) eq __PACKAGE__ &&
			ref($ip) eq __PACKAGE__;
  return (sub128($sref->{IP},$ip->{NT}) && sub128($ip->{BC},$sref->{IP}))
	? 1 : 0;
}

=item * $rv = $ip->contains($someip);

Check to see if $ip is within the cidr range of $someip. i.e.

This is the logical compliment of the B<within> method.

=cut

sub contains {
  my($sref,$ip) = @_;
  return within($ip,$sref);
}

=pod

=back

=head1 COPYRIGHT

Copyright 2006, Michael Robinton <michael@bizsystems.com>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License (except as noted
otherwise in individuals sub modules)  published by
the Free Software Foundation; either version 2 of the License, or 
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of 
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the  
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

=head1 AUTHOR

Michael Robinton <michael@bizsystems.com>

=head1 SEE ALSO

	man (3) NetAddr::IP::Util

=cut

1;


Makefile.PL
use ExtUtils::MakeMaker;
use Config;
use lib qw(inc);

my $pkg = 'Net::Connection::Sniffer';
$pkg =~ /[^:]+$/;
my $module = $& .'.pm';

my $view = $then + 86400 < $now ? 5:0;

if ($Config{osname} =~ /win/i) {
  print q|Windows and Darwin not supported
Volunteers welcome
|;
  exit 1;
}
  

unless (eval {
        require Sys::LoadAvg &&
	$Sys::LoadAvg::VERSION >= 0.3
	}) {
  print q|
optional Sys::LoadAvg v0.3 or greater is required if 
you want to quench excess CPU usage during 'coalesc'
operation.

|;
  sleep $view if $view;
}

# extra libs we would like to have that perl does not,
# a SPACE separated list
#
# put a test for them in configure.ac and let the script
# below find and insert them. Make sure and add the necessary
# paths in %makeparms for INC and LIBS but not the library itself
#
my $wish_libs = '-lpcap';

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

my %makeparms = (
    NAME		=> $pkg,
    VERSION_FROM	=> $module, # finds $VERSION
    PREREQ_PM		=> {
				Net::DNS::Codes		=> 0.09,
				Net::DNS::ToolKit	=> 0.42,
				Net::Interface		=> 1.009,
				Net::NBsocket		=> 0.14,
				Sys::Hostname::FQDN	=> 0.10,
				Proc::PidUtil		=> 0.08,
				NetAddr::IP		=> 4.026,
				Sys::Sig		=> 0.04,
				Net::NBsocket		=> 0.16,
			},
# LIBS list should be a single string to be compatible with script below
    LIBS		=> '-L/usr/local/lib',
    INC			=> '-I/usr/local/include',
    clean		=> { FILES    => "*~ tmp* auto*"},
    realclean		=> { FILES    => "config.h config.log config.status"},
    dist		=> {COMPRESS=>'gzip', SUFFIX=>'gz'}
);

################################################
# fix up compile and link flags for "configure"
################################################
my %addflags = (
	ldflags		=> $makeparms{LIBS},
	lddlflags	=> $makeparms{LIBS},
	ccflags		=> $makeparms{INC},
);

my $cfgstrg = '';
foreach (sort keys %addflags) {
  my $KEY = uc $_;
  my $oldstuff = $Config{$_} =~ /(\S.+)/ ? $1 : '';
  $oldstuff .= ' ' if $oldstuff && $oldstuff !~ / $/;
#print "$_, $KEY, $oldstuff, $addflags{$_}\n";
  unless ($oldstuff =~ m|$addflags{$_}|) {
    $oldstuff .= $addflags{$_};
  }
  $cfgstrg .= qq|$KEY="$oldstuff" |;
}

unless (-e './config.h') {
  my $command = qq|./configure $cfgstrg|;
  print $command,"\n";
  system($command);   
}

###################################
# fix up lib list
###################################

my %LIBS;
open(F,'config.h') or die "could not open config.h\n";
foreach(<F>) {
  if ($_ =~ /^#define LIBS([ a-zA-Z-]+)/) {
# make lib list unique
    map {$LIBS{$_} = 1} ($1 =~ /[a-zA-Z0-9-]+/g);
    last;
  }
}
close F;

my $liblist = $Config{libs} .' '. $wish_libs;

my $link = $makeparms{LIBS} =~ /(\S.+)/ ? $1 : '';
$link .= ' ' unless $link =~ / $/;
foreach(keys %LIBS) {
  if ($liblist =~ /$_\b/) {
    $link .= $_ .' ';   
  }
}
chop $link;
$makeparms{LIBS} = [$link];

###################################

sub MY::top_targets {
  package MY;
  my $inherited = shift->SUPER::top_targets(@_);
  $inherited =~ s/(pure_all\s+::.+)/$1 README/;
  $inherited;
}

sub MY::post_constants {
  my $post_constants = q|
MY_POD2TEXT = |. $Config{scriptdirexp} .'/pod2text' .q|   
|;
} 

sub MY::postamble {
  package MY;
  my $postamble = q|
README  : |. $module .q|
	@$(MY_POD2TEXT) |. $module .q| > README

|;
}

WriteMakefile(%makeparms);

unless (open(F,'./config.h')) {
  warn "can not find ./config.h\n";
  exit;
}

unless (scalar grep {/#define\s+HAVE_PCAP/} (<F>)) {
  close F;
  warn "prerequisite 'libpcap' (-lpcap) not found\n";
  exit 0;
}
close F;

